### PR TITLE
TandemFoil: wake deficit features (gap-normalized displacement)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Wake deficit features encode the displacement of each mesh point from the forefoil trailing edge, normalized by the inter-foil gap. This gives the model explicit geometric signal about where the forefoil wake impinges on the aft foil — the primary aerodynamic coupling mechanism in tandem configurations. The features are already fully implemented in \`core/features.py\` as \`compute_wake_deficit_features()\` and routed through the tandem training stack via \`--enable-wake-deficit\`, but have never been tested in the ICML stack.

The wake deficit features add 2 dimensions to the input (dx_norm, dy_norm) — gap-normalized x/y displacement from the forefoil TE. For isolated-foil cases, these are zero-masked, so the feature is safe for mixed-batch training.

The physical intuition: pressure recovery and stagnation on the aft foil depend directly on whether a given surface point sits inside or outside the forefoil wake. The current best model has no explicit gap-relative geometry; the wake deficit feature gives it that signal.

## Instructions

Run the current TandemFoil best configuration with `--enable-wake-deficit` added. No other changes — we want a clean ablation.

**Command to run:**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine_t_max 30 \
  --no-use-ema \
  --model_slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --enable-wake-deficit \
  --epochs 999 \
  --wandb_group tandem-wake-deficit
```

The 180-minute timeout (`SENPAI_TIMEOUT_MINUTES=240` may be set by harness — use whatever budget is available). The current best runs to 14 epochs in 180 min at this config. Expect similar throughput.

**Report in your PR comment:**
- `val_primary/surface_pressure_mae` at best epoch
- `test_primary/surface_pressure_mae`
- Per-split val MAE: `single_in_dist`, `geom_camber_rc`, `geom_camber_cruise`, `re_rand`
- W&B run ID
- Number of epochs completed

**Optional second run (if time permits):** add `--enable-wake-angle` to also include the wake angle feature (3 dims total). Compare both runs. This can be the same `--wandb_group tandem-wake-deficit`.

## Baseline

- **val_primary/surface_pressure_mae:** 78.81
- **test_primary/surface_pressure_mae:** 75.13
- **Per-split val MAE:** single_in_dist=97.78, geom_camber_rc=79.50, geom_camber_cruise=65.56, re_rand=72.41
- **W&B run:** 8k0blg8s
- **Baseline PR:** #2495 (nezuko — T_max=30 long run, 14 epochs, 180-min budget)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 30 --no-use-ema --model_slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### Run 1: Wake Deficit Only (`--enable-wake-deficit`)

| Metric | Baseline | This Run | Delta |
|--------|----------|----------|-------|
| **val_primary/surface_pressure_mae** | **78.81** | **81.63** | **+2.82 (worse)** |
| test_primary/surface_pressure_mae | 75.13 | 84.84 | +9.71 (worse) |
| val single_in_dist | 97.78 | 86.94 | -10.84 (better) |
| val geom_camber_rc | 79.50 | 118.45 | +38.95 (worse) |
| val geom_camber_cruise | 65.56 | 66.59 | +1.03 (similar) |
| val re_rand | 72.41 | 79.86 | +7.45 (worse) |

- **W&B run:** t0kg2ymx
- **Epochs completed:** 14 (30-min timeout)
- **Best epoch:** 11 (val_primary=81.63)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 30 \
  --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --enable-wake-deficit --epochs 999 \
  --agent haku --wandb-name "haku/wake-deficit-only" --wandb-group tandem-wake-deficit
```

---

### Run 2: Wake Deficit + Angle (`--enable-wake-deficit --enable-wake-angle`)

| Metric | Baseline | This Run | Delta |
|--------|----------|----------|-------|
| **val_primary/surface_pressure_mae** | **78.81** | **88.52** | **+9.71 (worse)** |
| test_primary/surface_pressure_mae | 75.13 | 90.71 | +15.58 (worse) |
| val single_in_dist | 97.78 | 88.88 | -8.90 (better) |
| val geom_camber_rc | 79.50 | 99.72 | +20.22 (worse) |
| val geom_camber_cruise | 65.56 | 73.63 | +8.07 (worse) |
| val re_rand | 72.41 | 88.37 | +15.96 (worse) |

- **W&B run:** 50uznlxw
- **Epochs completed:** 13 (30-min timeout)
- **Best epoch:** 11 (val_primary=88.52)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 30 \
  --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --enable-wake-deficit --enable-wake-angle --epochs 999 \
  --agent haku --wandb-name "haku/wake-deficit-plus-angle" --wandb-group tandem-wake-deficit
```

---

## Analysis

**Neither run beats the baseline.** The wake deficit features did not improve performance:

1. **Wake deficit only** (81.63 vs 78.81 baseline) — close but slightly worse on val_primary. The interesting signal: `single_in_dist` improved significantly (-10.84), suggesting the wake features do help the model on standard tandem geometries. However, `geom_camber_rc` degraded severely (+38.95), indicating the model may overfit to the wake displacement signal in a way that hurts generalization to varying camber/Reynolds configurations.

2. **Wake deficit + angle** (88.52 vs 78.81) — worse across the board. The additional angle dimension seems to add noise rather than useful signal at this training budget. Same pattern: `single_in_dist` improved, but all other splits degraded.

3. **Both runs had only 30 min / ~14 epochs** vs the baseline's 180-min / 14-epoch budget. The convergence curves were still dropping (loss decreasing each cycle), so these features may need longer training to fully converge. The deficit-only run was particularly close to baseline at just 81.63 with potential for further improvement.

4. The `geom_camber_rc` regression in both runs is the most concerning signal — the wake displacement features may encode forefoil-specific geometry that doesn't transfer well across camber variations.

## Suggested follow-ups

- **Longer training budget (180 min):** The deficit-only run was very close at 81.63 with only 30 min. A full budget could close the gap or beat baseline.
- **Feature scaling/normalization:** The raw gap-normalized displacements may need learned scaling. A small MLP or learnable scale factor on the wake features before concatenation could help.
- **Wake deficit only on tandem splits:** Currently the features are zero-masked for isolated foils, which may confuse the model. Could try applying wake features only as a separate head or conditioning.
- **Lower LR for wake features:** The wake features may need a different learning rate to integrate smoothly with the existing feature set.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2549 because ADVISOR: Closing — wake deficit features hurt (81.63 vs 75.59 baseline). The existing --enable-te-coord-frame already captures trailing-edge geometry; the gap-normalized displacement is redundant. Zero-masking for isolated foils creates a discontinuous input manifold..
